internal/pkg/siderolink/logstorage/logstorage.go

```diff
@@ -0,0 +1,299 @@
+// included in the LICENSE file.
+
+// Package logstorage implements storage for storing logs of machines on the filesystem.
+package logstorage
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/siderolabs/gen/optional"
+	"github.com/ulikunitz/xz"
+	"go.uber.org/zap"
+)
+
+// Storage stores logs for machines on file system.
+type Storage struct {
+	dirtyMachineIDs map[string]struct{}
+	logger          *zap.Logger
+	Path            string
+	Compress        bool
+	lock            sync.Mutex
+}
+
+// New creates a new Storage.
+func New(path string, compress bool, logger *zap.Logger) *Storage {
+	return &Storage{
+		Path:            path,
+		Compress:        compress,
+		dirtyMachineIDs: map[string]struct{}{},
+		logger:          logger,
+	}
+}
+
+// MarkDirty marks the machine with the given ID as dirty.
+func (l *Storage) MarkDirty(id string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.dirtyMachineIDs[id] = struct{}{}
+}
+
+// Exists returns true if the log file exists for the given machine ID.
+func (l *Storage) Exists(id string) (bool, error) {
+	_, err := os.Stat(l.logPath(id, l.Compress))
+	if err == nil {
+		return true, nil
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return false, err
+	}
+
+	_, err = os.Stat(l.logPath(id, !l.Compress))
+	if err == nil {
+		return true, nil
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return false, err
+	}
+
+	return false, nil
+}
+
+// Load loads logs of the machine with the given id, if exists, into the given writer.
+func (l *Storage) Load(id string, writer io.Writer) error {
+	compressed := l.Compress
+
+	bufferDataOptional, hashHexActual, err := l.readLogFile(id, l.Compress)
+	if err != nil {
+		return err
+	}
+
+	if !bufferDataOptional.IsPresent() {
+		if bufferDataOptional, hashHexActual, err = l.readLogFile(id, !l.Compress); err != nil {
+			return err
+		}
+
+		compressed = !compressed
+	}
+
+	bufferData, ok := bufferDataOptional.Get()
+	if !ok {
+		return nil
+	}
+
+	hashHexExpectedBytes, err := os.ReadFile(l.logSHA256Path(id, compressed))
+	if err != nil {
+		removeErr := l.Remove(id)
+		if removeErr != nil {
+			return fmt.Errorf("failed to delete invalid log buffer file: %w", removeErr)
+		}
+
+		return fmt.Errorf("failed to read log buffer hash file: %w", err)
+	}
+
+	hashHexExpected := string(hashHexExpectedBytes)
+
+	// verify the hash
+	if hashHexExpected != hashHexActual {
+		removeErr := l.Remove(id)
+		if removeErr != nil {
+			return fmt.Errorf("failed to delete invalid log buffer file: %w", removeErr)
+		}
+
+		return fmt.Errorf("invalid log buffer hash in file: expected %s, got %s", hashHexExpected, hashHexActual)
+	}
+
+	_, err = io.Copy(writer, bytes.NewReader(bufferData))
+
+	return err
+}
+
+func (l *Storage) readLogFile(id string, compressed bool) (optional.Optional[[]byte], string, error) {
```

Comment:
reading log files should support both compressed an uncompressed, otherwise we would not be able to make a switch

Reply:
Yep, already doing that at the call site of this: https://github.com/siderolabs/omni/pull/168/files#diff-1601a497b172d27a79fe9d4e7b1c6718d5269c89637b589edbeaa2ae155a5889R77-R90

I implemented the feature with the switch in mind, so turning it on and off (as many times as you want) works as expected - nothing gets lost.

Reply:
Hmmm... I think it's trivial to detect whether the file is compressed or not: https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#zstandard-frames

So we probably don't need to try to read it twice to understand that? 

We could also encode the compression format in our file format if we need to. 

I also wonder if we could compress logs in memory and/or flush them from memory, as we certainly eat some RAM by simply having (active) logs in RAM all the time. From this point of view, we might benefit from compressing files in chunks vs. trying to compress it all, or even having compressed/uncompressed chunks.

So if we go with compression today, what is our path to reduce RAM usage?

Reply:
> So we probably don't need to try to read it twice to understand that?
The compressed file has a different extension (`.zst` in this case), that's why we try to read it twice. I did not want to leave the extension as `.log` if it was compressed, as `.log` kinda hints plaintext.

So only one of `a.log` or `a.log.zst` would exist on the filesystem at any time, so we don't read anything twice, we just probe to find out which one is on the disk.

> We could also encode the compression format in our file format if we need to.
We could, but if we go with the extension, we don't need that.

> So if we go with compression today, what is our path to reduce RAM usage?
My aim was to mainly address the disk pressure problem we had recently. Compressing in-memory needs some more work as it would touch the current circular buffer model we are using. Should I also go for that, in this PR?

Reply:
> So only one of a.log or a.log.zst would exist on the filesystem at any time, so we don't read anything twice, we just probe to find out which one is on the disk.

what if we have both files? now I'm even more confused :exploding_head: 

> My aim was to mainly address the disk pressure problem we had recently. Compressing in-memory needs some more work as it would touch the current circular buffer model we are using. Should I also go for that, in this PR?

No, I'm not asking for that of course, but I want us to draw a plan what it might be if we go with that. So that our current implementation at least fits the future direction. E.g. if the circular buffer is already compressed, we don't need to compress on writes, and we don't need to decompress on reads. But certainly we can't compress it the way we do today as well.